### PR TITLE
allow directives on variable definitions

### PR DESCRIPTION
**Pull request by dugenkui03**
Load to #1765 , Allow directives on variable definitions.

### CODE REVIEW

File: src/main/java/graphql/schema/GraphQLDirective.java [hunk 1]
```diff
@@ -209,6 +224,7 @@ public Builder(GraphQLDirective existing) {
             this.onOperation = existing.isOnOperation();
             this.onFragment = existing.isOnFragment();
             this.onField = existing.isOnField();
+            this.onVariableDefinition=existing.isOnVariableDefinition();
```
Comment (andimarek) [inline src/main/java/graphql/schema/GraphQLDirective.java:227]: the formatting here is off a bit

File: src/main/java/graphql/schema/GraphQLDirective.java [hunk 2]
```diff
@@ -51,8 +52,9 @@ public GraphQLDirective(String name,
                             List<GraphQLArgument> arguments,
                             boolean onOperation,
                             boolean onFragment,
-                            boolean onField) {
-        this(name, description, locations, arguments, onOperation, onFragment, onField, null);
+                            boolean onField,
+                            boolean onVariableDefinition) {
+        this(name, description, locations, arguments, onOperation, onFragment, onField,onVariableDefinition,null);
```
Comment (andimarek) [inline src/main/java/graphql/schema/GraphQLDirective.java:57]: we should not add new constructors.

File: src/main/java/graphql/language/VariableDefinition.java [hunk 3]
```diff
@@ -50,7 +53,22 @@ protected VariableDefinition(String name,
...
+
+    /**
+     * alternative to using a Builder for convenience
+     *
+     * @param name         of the variable
+     * @param type         of the variable
+     * @param defaultValue of the variable
+     * @param directives   of the variable
+     */
+    public VariableDefinition(String name,
```
Comment (andimarek) [inline src/main/java/graphql/language/VariableDefinition.java:67]: we should not add any new constructors, but force new code to use the builder

File: src/main/java/graphql/schema/GraphQLDirective.java [hunk 4]
```diff
@@ -342,6 +358,19 @@ public Builder onField(boolean onField) {
             return this;
         }
 
+        /**
+         * @param onVariableDefinition onVariableDefinition
+         *
+         * @return this builder
+         *
+         * @deprecated Use {@code graphql.schema.GraphQLDirective.Builder#validLocations(DirectiveLocation...)}
```
Comment (andimarek) [inline src/main/java/graphql/schema/GraphQLDirective.java:366]: lets not add new methods which are immediately deprecated.

File: src/main/java/graphql/schema/GraphQLDirective.java (hunk 2 above)
Comment (dugenkui03) [reply src/main/java/graphql/schema/GraphQLDirective.java:57]: that is not a new constructor, but added new field(onVariableDefinition) in constructor

File: src/main/java/graphql/language/VariableDefinition.java (hunk 3 above)
Comment (dugenkui03) [reply src/main/java/graphql/language/VariableDefinition.java]: thanks for you advice, I have delete this new constructor.

File: src/main/java/graphql/schema/GraphQLDirective.java (hunk 4 above)
Comment (dugenkui03) [reply src/main/java/graphql/schema/GraphQLDirective.java]: copy that.

**Comment by dugenkui03** [comment]
> thanks a lot.
> In general it looks fine, but I left a few comments.
> Also: inside the graphql java project is an IntelliJ formatter which we use to format. If you use intellij please apply it.
> 
> Thanks!

Thanks for your review and advice. I have update IntelliJ setting and this PR. 

ps: `isOnVariableDefinition` in `GraphQLDirective` haven't removed, because it is used in some method, such as `Introspection` and `GraphQLDirective` and `KnownDirectives`. should I replace it with other methods.

**Comment by andimarek** [comment]
Thanks @dugenkui03: I removed a not needed boolean and merged it.

**Pull request merged**